src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/CoverageModelEMComputeBlock.java

```diff
@@ -0,0 +1,1332 @@
+import org.broadinstitute.hellbender.tools.coveragemodel.cachemanager.Duplicable;
+import org.broadinstitute.hellbender.tools.coveragemodel.cachemanager.DuplicableNDArray;
+import org.broadinstitute.hellbender.tools.coveragemodel.cachemanager.DuplicableNumber;
+import org.broadinstitute.hellbender.tools.coveragemodel.cachemanager.ImmutableComputableGraph;
+import org.broadinstitute.hellbender.tools.coveragemodel.math.RobustBrentSolver;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.param.ParamUtils;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.factory.Nd4j;
+import org.nd4j.linalg.indexing.NDArrayIndex;
+import org.nd4j.linalg.ops.transforms.Transforms;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * This class represents an immutable block of data containers, query methods and cloners corresponding to
+ * a partition of the target space.
+ *
+ * TODO github/gatk-protected issue #853 -- logging in spark mode (log4j is not serializable)
+ * TODO github/gatk-protected issue #853 -- use instrumentation to measure memory consumption
+
+ * @implNote Methods that manipulate INDArrays must make sure to leave queried values from
+ * {@link CoverageModelEMComputeBlock#icg} unchanged. For example, to calculate the A.B.C + D,
+ * the correct template is first fetch the data from ICG as follows:
+ *
+ * // query cached values
+ * final INDArray A = getINDArrayFromCache(CoverageModelICGCacheNode.A);
+ * final INDArray B = getINDArrayFromCache(CoverageModelICGCacheNode.B);
+ * final INDArray C = getINDArrayFromCache(CoverageModelICGCacheNode.C);
+ * final INDArray D = getINDArrayFromCache(CoverageModelICGCacheNode.D);
+ *
+ * and then call:
+ *
+ * // calculations
+ * final INDArray result_1 = A.mul(B).muli(C).addi(D); // OK
+ *                           ---1--- (a copy is instantiated; call it tmp)
+ *                                ----2---  (tmp is modified in-place)
+ *                                    -----3---- (tmp is modified in-place)
+ *
+ * (note the difference between mul(...) and muli(...)
+ *
+ * Here, A and multiplied by B to create a new INDArray; subsequent operations can be performed
+ * in-place to avoid unnecessary memory allocation/deallocation and garbage generation.
+ *
+ * It is, however, illegal to perform the following operation:
+ *
+ * final INDArray result_3 = A.muli(B).muli(C).addi(D); // NOT OK
+ *                           ---1--- (A is modified in-place!)
+ *
+ * Here, A is modified in-place which corrupts the data in {@link CoverageModelEMComputeBlock#icg}
+ *
+ * The code reviewer must CAREFULLY check all of the matrix computations to ensure that the such
```

Comment:
(I just did that and I got identical results -- at least we don't have to _worry_ for now, though, this doesn't obviate better engineering)